{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "613e6e14-4611-4387-b286-5d5e7b4d5b23",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import os\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "99d17d91-c2a8-45a4-9143-ec6b8552c443",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load PFC data\n",
    "adata = sc.read_h5ad('RNA_031324_PFC_13samples_82179cells.h5ad')\n",
    "\n",
    "#Separate out the metadata\n",
    "obs=adata.obs\n",
    "\n",
    "# Load sequencing metadata\n",
    "meta=pd.read_csv(\"SCORCH_tracking_metadata.tsv\",sep='\\t',\n",
    "                usecols=['Sample ID','Donor ID','Brain region', 'deep Seq ID (ATAC)', 'deep Seq ID (RNA)']\n",
    "                )\n",
    "meta=meta.dropna()\n",
    "meta=meta[meta['Brain region']=='PFC']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "26829119-52cf-438b-9d9b-7ff9942c65ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig.ident</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>percent.mt</th>\n",
       "      <th>nCount_RNA_raw</th>\n",
       "      <th>nFeature_RNA_raw</th>\n",
       "      <th>nCount_ATAC</th>\n",
       "      <th>nFeature_ATAC</th>\n",
       "      <th>gex_raw_reads</th>\n",
       "      <th>gex_mapped_reads</th>\n",
       "      <th>...</th>\n",
       "      <th>supertype_label</th>\n",
       "      <th>supertype_name</th>\n",
       "      <th>supertype_softmax_probability</th>\n",
       "      <th>leiden</th>\n",
       "      <th>donor</th>\n",
       "      <th>condition</th>\n",
       "      <th>age</th>\n",
       "      <th>sex</th>\n",
       "      <th>barcode</th>\n",
       "      <th>celltype_converted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CG103_PFC_hiv_AAACAGCCACAGCCTG-1</th>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>2068.0</td>\n",
       "      <td>1210</td>\n",
       "      <td>2.949710</td>\n",
       "      <td>2150.0</td>\n",
       "      <td>1260</td>\n",
       "      <td>1902.0</td>\n",
       "      <td>866</td>\n",
       "      <td>7730</td>\n",
       "      <td>7239</td>\n",
       "      <td>...</td>\n",
       "      <td>CS20230505_SUPT_0121</td>\n",
       "      <td>Oligo_2</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>CG103</td>\n",
       "      <td>HIV+OUD</td>\n",
       "      <td>40.2</td>\n",
       "      <td>Male</td>\n",
       "      <td>AAACAGCCACAGCCTG-1</td>\n",
       "      <td>Oligodendrocyte</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CG103_PFC_hiv_AAACAGCCACTAGCGT-1</th>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>2761.0</td>\n",
       "      <td>1473</td>\n",
       "      <td>0.144875</td>\n",
       "      <td>2886.0</td>\n",
       "      <td>1554</td>\n",
       "      <td>512.0</td>\n",
       "      <td>249</td>\n",
       "      <td>11192</td>\n",
       "      <td>10606</td>\n",
       "      <td>...</td>\n",
       "      <td>CS20230505_SUPT_0121</td>\n",
       "      <td>Oligo_2</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>CG103</td>\n",
       "      <td>HIV+OUD</td>\n",
       "      <td>40.2</td>\n",
       "      <td>Male</td>\n",
       "      <td>AAACAGCCACTAGCGT-1</td>\n",
       "      <td>Oligodendrocyte</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CG103_PFC_hiv_AAACAGCCATTGCGTA-1</th>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>26576.0</td>\n",
       "      <td>6515</td>\n",
       "      <td>0.594521</td>\n",
       "      <td>28381.0</td>\n",
       "      <td>7060</td>\n",
       "      <td>5538.0</td>\n",
       "      <td>2604</td>\n",
       "      <td>104039</td>\n",
       "      <td>98223</td>\n",
       "      <td>...</td>\n",
       "      <td>CS20230505_SUPT_0082</td>\n",
       "      <td>L5 IT_1</td>\n",
       "      <td>0.9958</td>\n",
       "      <td>8</td>\n",
       "      <td>CG103</td>\n",
       "      <td>HIV+OUD</td>\n",
       "      <td>40.2</td>\n",
       "      <td>Male</td>\n",
       "      <td>AAACAGCCATTGCGTA-1</td>\n",
       "      <td>L5_IT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CG103_PFC_hiv_AAACATGCAAGCTTAT-1</th>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>2587.0</td>\n",
       "      <td>1400</td>\n",
       "      <td>0.734441</td>\n",
       "      <td>2689.0</td>\n",
       "      <td>1463</td>\n",
       "      <td>261.0</td>\n",
       "      <td>121</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>...</td>\n",
       "      <td>CS20230505_SUPT_0121</td>\n",
       "      <td>Oligo_2</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>CG103</td>\n",
       "      <td>HIV+OUD</td>\n",
       "      <td>40.2</td>\n",
       "      <td>Male</td>\n",
       "      <td>AAACATGCAAGCTTAT-1</td>\n",
       "      <td>Oligodendrocyte</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CG103_PFC_hiv_AAACATGCACTTACAG-1</th>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>1004.0</td>\n",
       "      <td>682</td>\n",
       "      <td>0.199203</td>\n",
       "      <td>1030.0</td>\n",
       "      <td>702</td>\n",
       "      <td>407.0</td>\n",
       "      <td>198</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>...</td>\n",
       "      <td>CS20230505_SUPT_0121</td>\n",
       "      <td>Oligo_2</td>\n",
       "      <td>0.9999</td>\n",
       "      <td>0</td>\n",
       "      <td>CG103</td>\n",
       "      <td>HIV+OUD</td>\n",
       "      <td>40.2</td>\n",
       "      <td>Male</td>\n",
       "      <td>AAACATGCACTTACAG-1</td>\n",
       "      <td>Oligodendrocyte</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CE290_PFC_hiv_TTTGTGTTCAAACTCA-1</th>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>7098.0</td>\n",
       "      <td>2813</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7396.0</td>\n",
       "      <td>2986</td>\n",
       "      <td>5369.0</td>\n",
       "      <td>2538</td>\n",
       "      <td>16718</td>\n",
       "      <td>15851</td>\n",
       "      <td>...</td>\n",
       "      <td>CS20230505_SUPT_0049</td>\n",
       "      <td>Sst_22</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>4</td>\n",
       "      <td>CE290</td>\n",
       "      <td>OUD</td>\n",
       "      <td>35.0</td>\n",
       "      <td>Male</td>\n",
       "      <td>TTTGTGTTCAAACTCA-1</td>\n",
       "      <td>Sst</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CE290_PFC_hiv_TTTGTGTTCAATGACC-1</th>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>4619.0</td>\n",
       "      <td>2225</td>\n",
       "      <td>0.151548</td>\n",
       "      <td>4803.0</td>\n",
       "      <td>2317</td>\n",
       "      <td>5372.0</td>\n",
       "      <td>2494</td>\n",
       "      <td>11832</td>\n",
       "      <td>11276</td>\n",
       "      <td>...</td>\n",
       "      <td>CS20230505_SUPT_0046</td>\n",
       "      <td>Sst Chodl_2</td>\n",
       "      <td>0.9927</td>\n",
       "      <td>4</td>\n",
       "      <td>CE290</td>\n",
       "      <td>OUD</td>\n",
       "      <td>35.0</td>\n",
       "      <td>Male</td>\n",
       "      <td>TTTGTGTTCAATGACC-1</td>\n",
       "      <td>Sst_Chodl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CE290_PFC_hiv_TTTGTGTTCTAAGGAG-1</th>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>10573.0</td>\n",
       "      <td>4421</td>\n",
       "      <td>0.548567</td>\n",
       "      <td>11320.0</td>\n",
       "      <td>4720</td>\n",
       "      <td>16483.0</td>\n",
       "      <td>7095</td>\n",
       "      <td>30376</td>\n",
       "      <td>29552</td>\n",
       "      <td>...</td>\n",
       "      <td>CS20230505_SUPT_0079</td>\n",
       "      <td>L4 IT_2</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>9</td>\n",
       "      <td>CE290</td>\n",
       "      <td>OUD</td>\n",
       "      <td>35.0</td>\n",
       "      <td>Male</td>\n",
       "      <td>TTTGTGTTCTAAGGAG-1</td>\n",
       "      <td>L4_IT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CE290_PFC_hiv_TTTGTTGGTAACGAGG-1</th>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>8148.0</td>\n",
       "      <td>3636</td>\n",
       "      <td>0.049092</td>\n",
       "      <td>8587.0</td>\n",
       "      <td>3828</td>\n",
       "      <td>1310.0</td>\n",
       "      <td>672</td>\n",
       "      <td>23237</td>\n",
       "      <td>22341</td>\n",
       "      <td>...</td>\n",
       "      <td>CS20230505_SUPT_0070</td>\n",
       "      <td>L2/3 IT_6</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>3</td>\n",
       "      <td>CE290</td>\n",
       "      <td>OUD</td>\n",
       "      <td>35.0</td>\n",
       "      <td>Male</td>\n",
       "      <td>TTTGTTGGTAACGAGG-1</td>\n",
       "      <td>L2-3_IT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CE290_PFC_hiv_TTTGTTGGTTTAGCGA-1</th>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>12668.0</td>\n",
       "      <td>4810</td>\n",
       "      <td>0.797284</td>\n",
       "      <td>13525.0</td>\n",
       "      <td>5155</td>\n",
       "      <td>11235.0</td>\n",
       "      <td>5084</td>\n",
       "      <td>36980</td>\n",
       "      <td>35802</td>\n",
       "      <td>...</td>\n",
       "      <td>CS20230505_SUPT_0079</td>\n",
       "      <td>L4 IT_2</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>9</td>\n",
       "      <td>CE290</td>\n",
       "      <td>OUD</td>\n",
       "      <td>35.0</td>\n",
       "      <td>Male</td>\n",
       "      <td>TTTGTTGGTTTAGCGA-1</td>\n",
       "      <td>L4_IT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>82179 rows × 69 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     orig.ident  nCount_RNA  nFeature_RNA  \\\n",
       "CG103_PFC_hiv_AAACAGCCACAGCCTG-1  SeuratProject      2068.0          1210   \n",
       "CG103_PFC_hiv_AAACAGCCACTAGCGT-1  SeuratProject      2761.0          1473   \n",
       "CG103_PFC_hiv_AAACAGCCATTGCGTA-1  SeuratProject     26576.0          6515   \n",
       "CG103_PFC_hiv_AAACATGCAAGCTTAT-1  SeuratProject      2587.0          1400   \n",
       "CG103_PFC_hiv_AAACATGCACTTACAG-1  SeuratProject      1004.0           682   \n",
       "...                                         ...         ...           ...   \n",
       "CE290_PFC_hiv_TTTGTGTTCAAACTCA-1  SeuratProject      7098.0          2813   \n",
       "CE290_PFC_hiv_TTTGTGTTCAATGACC-1  SeuratProject      4619.0          2225   \n",
       "CE290_PFC_hiv_TTTGTGTTCTAAGGAG-1  SeuratProject     10573.0          4421   \n",
       "CE290_PFC_hiv_TTTGTTGGTAACGAGG-1  SeuratProject      8148.0          3636   \n",
       "CE290_PFC_hiv_TTTGTTGGTTTAGCGA-1  SeuratProject     12668.0          4810   \n",
       "\n",
       "                                  percent.mt  nCount_RNA_raw  \\\n",
       "CG103_PFC_hiv_AAACAGCCACAGCCTG-1    2.949710          2150.0   \n",
       "CG103_PFC_hiv_AAACAGCCACTAGCGT-1    0.144875          2886.0   \n",
       "CG103_PFC_hiv_AAACAGCCATTGCGTA-1    0.594521         28381.0   \n",
       "CG103_PFC_hiv_AAACATGCAAGCTTAT-1    0.734441          2689.0   \n",
       "CG103_PFC_hiv_AAACATGCACTTACAG-1    0.199203          1030.0   \n",
       "...                                      ...             ...   \n",
       "CE290_PFC_hiv_TTTGTGTTCAAACTCA-1    0.000000          7396.0   \n",
       "CE290_PFC_hiv_TTTGTGTTCAATGACC-1    0.151548          4803.0   \n",
       "CE290_PFC_hiv_TTTGTGTTCTAAGGAG-1    0.548567         11320.0   \n",
       "CE290_PFC_hiv_TTTGTTGGTAACGAGG-1    0.049092          8587.0   \n",
       "CE290_PFC_hiv_TTTGTTGGTTTAGCGA-1    0.797284         13525.0   \n",
       "\n",
       "                                  nFeature_RNA_raw  nCount_ATAC  \\\n",
       "CG103_PFC_hiv_AAACAGCCACAGCCTG-1              1260       1902.0   \n",
       "CG103_PFC_hiv_AAACAGCCACTAGCGT-1              1554        512.0   \n",
       "CG103_PFC_hiv_AAACAGCCATTGCGTA-1              7060       5538.0   \n",
       "CG103_PFC_hiv_AAACATGCAAGCTTAT-1              1463        261.0   \n",
       "CG103_PFC_hiv_AAACATGCACTTACAG-1               702        407.0   \n",
       "...                                            ...          ...   \n",
       "CE290_PFC_hiv_TTTGTGTTCAAACTCA-1              2986       5369.0   \n",
       "CE290_PFC_hiv_TTTGTGTTCAATGACC-1              2317       5372.0   \n",
       "CE290_PFC_hiv_TTTGTGTTCTAAGGAG-1              4720      16483.0   \n",
       "CE290_PFC_hiv_TTTGTTGGTAACGAGG-1              3828       1310.0   \n",
       "CE290_PFC_hiv_TTTGTTGGTTTAGCGA-1              5155      11235.0   \n",
       "\n",
       "                                  nFeature_ATAC  gex_raw_reads  \\\n",
       "CG103_PFC_hiv_AAACAGCCACAGCCTG-1            866           7730   \n",
       "CG103_PFC_hiv_AAACAGCCACTAGCGT-1            249          11192   \n",
       "CG103_PFC_hiv_AAACAGCCATTGCGTA-1           2604         104039   \n",
       "CG103_PFC_hiv_AAACATGCAAGCTTAT-1            121    -2147483648   \n",
       "CG103_PFC_hiv_AAACATGCACTTACAG-1            198    -2147483648   \n",
       "...                                         ...            ...   \n",
       "CE290_PFC_hiv_TTTGTGTTCAAACTCA-1           2538          16718   \n",
       "CE290_PFC_hiv_TTTGTGTTCAATGACC-1           2494          11832   \n",
       "CE290_PFC_hiv_TTTGTGTTCTAAGGAG-1           7095          30376   \n",
       "CE290_PFC_hiv_TTTGTTGGTAACGAGG-1            672          23237   \n",
       "CE290_PFC_hiv_TTTGTTGGTTTAGCGA-1           5084          36980   \n",
       "\n",
       "                                  gex_mapped_reads  ...       supertype_label  \\\n",
       "CG103_PFC_hiv_AAACAGCCACAGCCTG-1              7239  ...  CS20230505_SUPT_0121   \n",
       "CG103_PFC_hiv_AAACAGCCACTAGCGT-1             10606  ...  CS20230505_SUPT_0121   \n",
       "CG103_PFC_hiv_AAACAGCCATTGCGTA-1             98223  ...  CS20230505_SUPT_0082   \n",
       "CG103_PFC_hiv_AAACATGCAAGCTTAT-1       -2147483648  ...  CS20230505_SUPT_0121   \n",
       "CG103_PFC_hiv_AAACATGCACTTACAG-1       -2147483648  ...  CS20230505_SUPT_0121   \n",
       "...                                            ...  ...                   ...   \n",
       "CE290_PFC_hiv_TTTGTGTTCAAACTCA-1             15851  ...  CS20230505_SUPT_0049   \n",
       "CE290_PFC_hiv_TTTGTGTTCAATGACC-1             11276  ...  CS20230505_SUPT_0046   \n",
       "CE290_PFC_hiv_TTTGTGTTCTAAGGAG-1             29552  ...  CS20230505_SUPT_0079   \n",
       "CE290_PFC_hiv_TTTGTTGGTAACGAGG-1             22341  ...  CS20230505_SUPT_0070   \n",
       "CE290_PFC_hiv_TTTGTTGGTTTAGCGA-1             35802  ...  CS20230505_SUPT_0079   \n",
       "\n",
       "                                  supertype_name  \\\n",
       "CG103_PFC_hiv_AAACAGCCACAGCCTG-1         Oligo_2   \n",
       "CG103_PFC_hiv_AAACAGCCACTAGCGT-1         Oligo_2   \n",
       "CG103_PFC_hiv_AAACAGCCATTGCGTA-1         L5 IT_1   \n",
       "CG103_PFC_hiv_AAACATGCAAGCTTAT-1         Oligo_2   \n",
       "CG103_PFC_hiv_AAACATGCACTTACAG-1         Oligo_2   \n",
       "...                                          ...   \n",
       "CE290_PFC_hiv_TTTGTGTTCAAACTCA-1          Sst_22   \n",
       "CE290_PFC_hiv_TTTGTGTTCAATGACC-1     Sst Chodl_2   \n",
       "CE290_PFC_hiv_TTTGTGTTCTAAGGAG-1         L4 IT_2   \n",
       "CE290_PFC_hiv_TTTGTTGGTAACGAGG-1       L2/3 IT_6   \n",
       "CE290_PFC_hiv_TTTGTTGGTTTAGCGA-1         L4 IT_2   \n",
       "\n",
       "                                  supertype_softmax_probability  leiden  \\\n",
       "CG103_PFC_hiv_AAACAGCCACAGCCTG-1                         1.0000       0   \n",
       "CG103_PFC_hiv_AAACAGCCACTAGCGT-1                         1.0000       0   \n",
       "CG103_PFC_hiv_AAACAGCCATTGCGTA-1                         0.9958       8   \n",
       "CG103_PFC_hiv_AAACATGCAAGCTTAT-1                         1.0000       0   \n",
       "CG103_PFC_hiv_AAACATGCACTTACAG-1                         0.9999       0   \n",
       "...                                                         ...     ...   \n",
       "CE290_PFC_hiv_TTTGTGTTCAAACTCA-1                         1.0000       4   \n",
       "CE290_PFC_hiv_TTTGTGTTCAATGACC-1                         0.9927       4   \n",
       "CE290_PFC_hiv_TTTGTGTTCTAAGGAG-1                         1.0000       9   \n",
       "CE290_PFC_hiv_TTTGTTGGTAACGAGG-1                         1.0000       3   \n",
       "CE290_PFC_hiv_TTTGTTGGTTTAGCGA-1                         1.0000       9   \n",
       "\n",
       "                                  donor  condition   age   sex  \\\n",
       "CG103_PFC_hiv_AAACAGCCACAGCCTG-1  CG103    HIV+OUD  40.2  Male   \n",
       "CG103_PFC_hiv_AAACAGCCACTAGCGT-1  CG103    HIV+OUD  40.2  Male   \n",
       "CG103_PFC_hiv_AAACAGCCATTGCGTA-1  CG103    HIV+OUD  40.2  Male   \n",
       "CG103_PFC_hiv_AAACATGCAAGCTTAT-1  CG103    HIV+OUD  40.2  Male   \n",
       "CG103_PFC_hiv_AAACATGCACTTACAG-1  CG103    HIV+OUD  40.2  Male   \n",
       "...                                 ...        ...   ...   ...   \n",
       "CE290_PFC_hiv_TTTGTGTTCAAACTCA-1  CE290        OUD  35.0  Male   \n",
       "CE290_PFC_hiv_TTTGTGTTCAATGACC-1  CE290        OUD  35.0  Male   \n",
       "CE290_PFC_hiv_TTTGTGTTCTAAGGAG-1  CE290        OUD  35.0  Male   \n",
       "CE290_PFC_hiv_TTTGTTGGTAACGAGG-1  CE290        OUD  35.0  Male   \n",
       "CE290_PFC_hiv_TTTGTTGGTTTAGCGA-1  CE290        OUD  35.0  Male   \n",
       "\n",
       "                                             barcode  celltype_converted  \n",
       "CG103_PFC_hiv_AAACAGCCACAGCCTG-1  AAACAGCCACAGCCTG-1     Oligodendrocyte  \n",
       "CG103_PFC_hiv_AAACAGCCACTAGCGT-1  AAACAGCCACTAGCGT-1     Oligodendrocyte  \n",
       "CG103_PFC_hiv_AAACAGCCATTGCGTA-1  AAACAGCCATTGCGTA-1               L5_IT  \n",
       "CG103_PFC_hiv_AAACATGCAAGCTTAT-1  AAACATGCAAGCTTAT-1     Oligodendrocyte  \n",
       "CG103_PFC_hiv_AAACATGCACTTACAG-1  AAACATGCACTTACAG-1     Oligodendrocyte  \n",
       "...                                              ...                 ...  \n",
       "CE290_PFC_hiv_TTTGTGTTCAAACTCA-1  TTTGTGTTCAAACTCA-1                 Sst  \n",
       "CE290_PFC_hiv_TTTGTGTTCAATGACC-1  TTTGTGTTCAATGACC-1           Sst_Chodl  \n",
       "CE290_PFC_hiv_TTTGTGTTCTAAGGAG-1  TTTGTGTTCTAAGGAG-1               L4_IT  \n",
       "CE290_PFC_hiv_TTTGTTGGTAACGAGG-1  TTTGTTGGTAACGAGG-1             L2-3_IT  \n",
       "CE290_PFC_hiv_TTTGTTGGTTTAGCGA-1  TTTGTTGGTTTAGCGA-1               L4_IT  \n",
       "\n",
       "[82179 rows x 69 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create barcode column so that we can add cell type data  \n",
    "obs['barcode']=[s.split('_')[-1] for s in obs.index.values]\n",
    "obs['celltype_converted']=obs['subclass_name'].str.replace('/','-').str.replace(' ','_')\n",
    "obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4b6f07fb-29ad-4be5-9a90-1e0f3267ad95",
   "metadata": {},
   "outputs": [],
   "source": [
    "cfe=pd.read_csv('CFE_directories.txt',delim_whitespace=True,\n",
    "               names=['folder','date','time','-'])\n",
    "\n",
    "cfe['folder']=cfe['folder'].str.replace('/','')\n",
    "# cfe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2415f61b-77ba-4588-8534-a533992f6b8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr_list = ['chr1', 'chr10', 'chr11', 'chr12','chr13', 'chr14', 'chr15', 'chr16', 'chr17', 'chr18', \n",
    "             'chr19', 'chr2', 'chr20', 'chr21', 'chr22', 'chr3', 'chr4',\n",
    "                'chr5', 'chr6', 'chr7', 'chr8', 'chr9', 'chrX', 'chrY']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "133209e0-cce6-446d-be32-7007be5ca928",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Oligodendrocyte',\n",
       " 'L5 IT',\n",
       " 'L2/3 IT',\n",
       " 'Vip',\n",
       " 'Pvalb',\n",
       " 'L6 IT Car3',\n",
       " 'L4 IT',\n",
       " 'Lamp5',\n",
       " 'Astrocyte',\n",
       " 'Lamp5 Lhx6',\n",
       " 'Microglia-PVM',\n",
       " 'OPC',\n",
       " 'L5/6 NP',\n",
       " 'Sncg',\n",
       " 'Sst',\n",
       " 'L6b',\n",
       " 'L6 IT',\n",
       " 'VLMC',\n",
       " 'L6 CT',\n",
       " 'Pax6',\n",
       " 'Endothelial',\n",
       " 'Chandelier',\n",
       " 'L5 ET',\n",
       " 'Sst Chodl']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(obs.subclass_name.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "4a38124a-a7ff-48b8-b5d0-8ef63eaeddd4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_bigwig_files(libid):\n",
    "# for libid in tqdm(meta['deep Seq ID (ATAC)']):\n",
    "    # Save donor id \n",
    "    donor_id = meta[meta['deep Seq ID (ATAC)']==libid]['Donor ID'].iloc[0]\n",
    "    print(donor_id)\n",
    "\n",
    "    # Save library id \n",
    "    donor=libid\n",
    "    # print(libid)\n",
    "    \n",
    "    split = (libid.split('_'))[:2]\n",
    "    alt_id = split[0] +'_' +split[1]+'_1_2'\n",
    "    # print(alt_id)\n",
    "\n",
    "    #Use combined file if it exists\n",
    "    myfolder=cfe[cfe['folder'].str.contains(alt_id)]\n",
    "\n",
    "    # if not, use the _2 file \n",
    "    if myfolder.shape[0] == 0:\n",
    "        myfolder=cfe[cfe['folder'].str.contains(libid)]\n",
    "    # print(myfolder.shape[0])\n",
    "\n",
    "    url='https://epigenomics.sdsc.edu/hjiao/bot_output/Tariq_Rana_415879/counts/'+myfolder['folder'].values\n",
    "    url=url[0]\n",
    "\n",
    "    # Download atac data from folder to scratch on server\n",
    "    if not os.path.exists(f'/scratch/{donor}.atac_fragments.tsv.gz'): \n",
    "        !wget {url}/outs/atac_fragments.tsv.gz -O /scratch/{donor}.atac_fragments.tsv.gz   \n",
    "        print('Data download successful...')\n",
    "\n",
    "    frags=pd.read_csv(f'/scratch/{donor}.atac_fragments.tsv.gz',sep='\\t',\n",
    "                # nrows=100000,\n",
    "                comment='#',\n",
    "                names=['chr','start','end','barcode','-'],\n",
    "                 )\n",
    "\n",
    "    for celltype in list(obs.subclass_name.unique()):\n",
    "        # Create savable file name for cell type\n",
    "        celltype_converted=celltype.replace('/','-').replace(' ','_')\n",
    "        # Subset dataframe to that of the donor and one cell type\n",
    "        my_obs=obs.loc[(obs['donor']==donor_id) & (obs['subclass_name']==celltype)]\n",
    "\n",
    "        #Keep only cells that are in the obs metadata frame\n",
    "        my_frags=frags[frags['barcode'].isin(my_obs['barcode'])]\n",
    "\n",
    "        # Keep only chromosomes 1-22, X and Y\n",
    "        my_frags = my_frags[my_frags['chr'].isin(chr_list)]\n",
    "\n",
    "        # Record atac fragment number so can normalize the file later\n",
    "        scale_num = (1/(my_frags.shape[0])) * 1e6\n",
    "\n",
    "        # Save to a bed file. Keep only first 3 columns of fragment file\n",
    "        bedfile=f'/scratch/{donor}.{celltype_converted}.atac_fragments.bed'\n",
    "        my_frags.iloc[:,:3].to_csv(bedfile,sep='\\t',index=False,header=False)\n",
    "\n",
    "        # Convert bed to bedgraph\n",
    "        bgfile=bedfile.replace('.bed','.bg')\n",
    "        genome='hg38_hiv1.genome'\n",
    "        print('Conversion to bedgraph successful for ' +celltype+'...')\n",
    "\n",
    "        # Get genome coverage\n",
    "        !bedtools genomecov -i {bedfile} -g {genome} -bg -scale {scale_num} > {bgfile}\n",
    "\n",
    "        # Create new file name\n",
    "        bigwig=bgfile.replace('.bg','.bw')\n",
    "        #Create bigwig file from bedgraph file\n",
    "        cmd=f'/cndd/bin/bedGraphToBigWig {bgfile} {genome} {bigwig}'\n",
    "        os.system(cmd)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8fc6dd8c-1144-4034-8dc9-49b6b5fe9113",
   "metadata": {},
   "outputs": [],
   "source": [
    "from multiprocessing import Pool"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b224104-613d-4367-8730-7623e2b31f24",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "0it [00:00, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CG103CA180CA482CE285CA650CE288CG118CE283CE286CC108CA600CA546CG130\n",
      "CE290\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "--2024-04-04 14:20:01--  https://epigenomics.sdsc.edu/hjiao/bot_output/Tariq_Rana_415879/counts/JB_1176_2_JB_1168_2/outs/atac_fragments.tsv.gz\n",
      "Resolving webproxy.ucsd.edu (webproxy.ucsd.edu)... 132.239.1.231, 132.239.1.230\n",
      "Connecting to webproxy.ucsd.edu (webproxy.ucsd.edu)|132.239.1.231|:3128... connected.\n",
      "Proxy request sent, awaiting response... 200 OK\n",
      "Length: 1970931793 (1.8G) [application/octet-stream]\n",
      "Saving to: ‘/scratch/JB_1168_2.atac_fragments.tsv.gz’\n",
      "\n",
      "/scratch/JB_1168_2. 100%[===================>]   1.83G   172MB/s    in 11s     \n",
      "\n",
      "2024-04-04 14:20:12 (173 MB/s) - ‘/scratch/JB_1168_2.atac_fragments.tsv.gz’ saved [1970931793/1970931793]\n",
      "\n",
      "Data download successful...\n",
      "Conversion to bedgraph successful for Oligodendrocyte...\n",
      "Conversion to bedgraph successful for Oligodendrocyte...\n",
      "Conversion to bedgraph successful for Oligodendrocyte...\n",
      "Conversion to bedgraph successful for Oligodendrocyte...\n",
      "Conversion to bedgraph successful for Oligodendrocyte...\n",
      "Conversion to bedgraph successful for Oligodendrocyte...\n",
      "Conversion to bedgraph successful for Oligodendrocyte...\n",
      "Conversion to bedgraph successful for L5 IT...\n",
      "Conversion to bedgraph successful for L5 IT...\n",
      "Conversion to bedgraph successful for L5 IT...\n",
      "Conversion to bedgraph successful for L5 IT...\n",
      "Conversion to bedgraph successful for L5 IT...\n",
      "Conversion to bedgraph successful for L5 IT...\n",
      "Conversion to bedgraph successful for L2/3 IT...\n",
      "Conversion to bedgraph successful for L2/3 IT...\n",
      "Conversion to bedgraph successful for L2/3 IT...\n",
      "Conversion to bedgraph successful for L2/3 IT...\n",
      "Conversion to bedgraph successful for L5 IT...\n",
      "Conversion to bedgraph successful for Vip...\n",
      "Conversion to bedgraph successful for L2/3 IT...\n",
      "Conversion to bedgraph successful for L2/3 IT...\n",
      "Conversion to bedgraph successful for Pvalb...\n",
      "Conversion to bedgraph successful for L6 IT Car3...\n",
      "Conversion to bedgraph successful for L4 IT...\n",
      "Conversion to bedgraph successful for L2/3 IT...\n",
      "Conversion to bedgraph successful for Vip...\n",
      "Conversion to bedgraph successful for Lamp5...\n",
      "Conversion to bedgraph successful for Pvalb...\n",
      "Conversion to bedgraph successful for Astrocyte...\n",
      "Conversion to bedgraph successful for Vip...\n",
      "Conversion to bedgraph successful for L6 IT Car3...\n",
      "Conversion to bedgraph successful for Pvalb...\n",
      "Conversion to bedgraph successful for Vip...\n",
      "Conversion to bedgraph successful for Vip...\n",
      "Conversion to bedgraph successful for L4 IT...\n",
      "Conversion to bedgraph successful for Lamp5 Lhx6...\n",
      "Conversion to bedgraph successful for L6 IT Car3...\n",
      "Conversion to bedgraph successful for Vip...\n",
      "Conversion to bedgraph successful for Pvalb...\n",
      "Conversion to bedgraph successful for Microglia-PVM...\n",
      "Conversion to bedgraph successful for Vip...\n",
      "Conversion to bedgraph successful for Pvalb...\n",
      "Conversion to bedgraph successful for Pvalb...\n",
      "Conversion to bedgraph successful for OPC...\n",
      "Conversion to bedgraph successful for L4 IT...\n",
      "Conversion to bedgraph successful for L6 IT Car3...\n",
      "Conversion to bedgraph successful for Lamp5...\n",
      "Conversion to bedgraph successful for L6 IT Car3...\n",
      "Conversion to bedgraph successful for L6 IT Car3...\n",
      "Conversion to bedgraph successful for Pvalb...\n",
      "Conversion to bedgraph successful for L5/6 NP...\n",
      "Conversion to bedgraph successful for L4 IT...\n",
      "Conversion to bedgraph successful for Astrocyte...\n",
      "Conversion to bedgraph successful for L4 IT...\n",
      "Conversion to bedgraph successful for Sncg...\n",
      "Conversion to bedgraph successful for L4 IT...\n",
      "Conversion to bedgraph successful for Sst...\n",
      "Conversion to bedgraph successful for Lamp5...\n",
      "Conversion to bedgraph successful for L6 IT Car3...\n",
      "Conversion to bedgraph successful for L6b...\n",
      "Conversion to bedgraph successful for Lamp5...\n",
      "Conversion to bedgraph successful for L4 IT...\n",
      "Conversion to bedgraph successful for Lamp5 Lhx6...\n",
      "Conversion to bedgraph successful for Astrocyte...\n",
      "Conversion to bedgraph successful for Lamp5...\n",
      "Conversion to bedgraph successful for Astrocyte...\n",
      "Conversion to bedgraph successful for Lamp5...\n",
      "Conversion to bedgraph successful for Microglia-PVM...\n",
      "Conversion to bedgraph successful for L6 IT...\n",
      "Conversion to bedgraph successful for Astrocyte...\n",
      "Conversion to bedgraph successful for OPC...\n",
      "Conversion to bedgraph successful for Astrocyte...\n",
      "Conversion to bedgraph successful for Lamp5 Lhx6...\n",
      "Conversion to bedgraph successful for Lamp5...\n",
      "Conversion to bedgraph successful for L5/6 NP...\n",
      "Conversion to bedgraph successful for VLMC...\n",
      "Conversion to bedgraph successful for Microglia-PVM...\n",
      "Conversion to bedgraph successful for L6 CT...\n",
      "Conversion to bedgraph successful for Sncg...\n",
      "Conversion to bedgraph successful for OPC...\n",
      "Conversion to bedgraph successful for Lamp5 Lhx6...\n",
      "Conversion to bedgraph successful for Lamp5 Lhx6...\n",
      "Conversion to bedgraph successful for Astrocyte...\n",
      "Conversion to bedgraph successful for Sst...\n",
      "Conversion to bedgraph successful for Pax6...\n",
      "Conversion to bedgraph successful for L5/6 NP...\n",
      "Conversion to bedgraph successful for Lamp5 Lhx6...\n",
      "Conversion to bedgraph successful for Microglia-PVM...\n",
      "Conversion to bedgraph successful for Microglia-PVM...\n",
      "Conversion to bedgraph successful for Endothelial...\n",
      "Conversion to bedgraph successful for Chandelier...\n",
      "Conversion to bedgraph successful for Sncg...\n",
      "Conversion to bedgraph successful for Microglia-PVM...\n",
      "Conversion to bedgraph successful for L6b...\n",
      "Conversion to bedgraph successful for OPC...\n",
      "Conversion to bedgraph successful for L5 ET...\n",
      "Conversion to bedgraph successful for Sst...\n",
      "Conversion to bedgraph successful for L6 IT...\n",
      "Conversion to bedgraph successful for OPC...\n",
      "Conversion to bedgraph successful for L6b...\n",
      "Conversion to bedgraph successful for OPC...\n",
      "Conversion to bedgraph successful for L5/6 NP...\n",
      "Conversion to bedgraph successful for L6 IT...\n",
      "Conversion to bedgraph successful for VLMC...\n",
      "Conversion to bedgraph successful for L5/6 NP...\n",
      "Conversion to bedgraph successful for Sncg...\n",
      "Conversion to bedgraph successful for Lamp5 Lhx6...\n",
      "Conversion to bedgraph successful for L6 CT...\n",
      "Conversion to bedgraph successful for VLMC...\n",
      "Conversion to bedgraph successful for Sncg...\n",
      "Conversion to bedgraph successful for L5/6 NP...\n",
      "Conversion to bedgraph successful for Pax6...\n",
      "Conversion to bedgraph successful for L6 CT...\n",
      "Conversion to bedgraph successful for Microglia-PVM...\n",
      "Conversion to bedgraph successful for Sst...\n",
      "Conversion to bedgraph successful for Sst...\n",
      "Conversion to bedgraph successful for Endothelial...\n",
      "Conversion to bedgraph successful for Sncg...\n",
      "Conversion to bedgraph successful for Pax6...\n",
      "Conversion to bedgraph successful for Chandelier...\n",
      "Conversion to bedgraph successful for OPC...\n",
      "Conversion to bedgraph successful for Sst...\n",
      "Conversion to bedgraph successful for Endothelial...\n",
      "Conversion to bedgraph successful for L6b...\n",
      "Conversion to bedgraph successful for L5 ET...\n",
      "Conversion to bedgraph successful for L6b...\n",
      "Conversion to bedgraph successful for Sst Chodl...\n",
      "Conversion to bedgraph successful for Chandelier...\n",
      "Conversion to bedgraph successful for L6 IT...\n",
      "Conversion to bedgraph successful for L6b...\n",
      "Conversion to bedgraph successful for L5/6 NP...\n",
      "Conversion to bedgraph successful for L6 IT...\n",
      "Conversion to bedgraph successful for L5 ET...\n",
      "Conversion to bedgraph successful for VLMC...\n",
      "Conversion to bedgraph successful for Sncg...\n",
      "Conversion to bedgraph successful for L6 IT...\n",
      "Conversion to bedgraph successful for VLMC...\n",
      "Conversion to bedgraph successful for L6 CT...\n",
      "Conversion to bedgraph successful for Sst...\n",
      "Conversion to bedgraph successful for L6 CT...\n",
      "Conversion to bedgraph successful for Pax6...\n",
      "Conversion to bedgraph successful for VLMC...\n",
      "Conversion to bedgraph successful for Pax6...\n",
      "Conversion to bedgraph successful for Endothelial...\n",
      "Conversion to bedgraph successful for L6 CT...\n",
      "Conversion to bedgraph successful for Endothelial...\n",
      "Conversion to bedgraph successful for L6b...\n",
      "Conversion to bedgraph successful for Chandelier...\n",
      "Conversion to bedgraph successful for Chandelier...\n",
      "Conversion to bedgraph successful for L6 IT...\n",
      "Conversion to bedgraph successful for Pax6...\n",
      "Conversion to bedgraph successful for L5 ET...\n",
      "Conversion to bedgraph successful for L5 ET...\n",
      "Conversion to bedgraph successful for Endothelial...\n",
      "Conversion to bedgraph successful for Sst Chodl...\n",
      "Conversion to bedgraph successful for VLMC...\n",
      "Conversion to bedgraph successful for Sst Chodl...\n",
      "Conversion to bedgraph successful for Chandelier...\n",
      "Conversion to bedgraph successful for L6 CT...\n",
      "Conversion to bedgraph successful for Pax6...\n",
      "Conversion to bedgraph successful for Endothelial...\n",
      "Conversion to bedgraph successful for Chandelier...\n",
      "Conversion to bedgraph successful for L5 ET...\n",
      "Conversion to bedgraph successful for Sst Chodl...\n"
     ]
    }
   ],
   "source": [
    "with Pool() as p:\n",
    "    list(tqdm(p.imap(create_bigwig_files, meta['deep Seq ID (ATAC)'])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8eff8f46-7c81-4016-8b73-0847ba1be9c3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
